File: dotnet/Falcon.PluginAPI/World/Level.cs
using System;
using System.Collections.Generic;
using Falcon.Interop;

namespace Falcon
{
    public sealed unsafe class Level : IEquatable<Level>
    {
        private readonly FalconLevel* _handle;

        internal Level(FalconLevel* handle)
        {
            _handle = handle;
        }

        public nint Handle
        {
            get
            {
                return (nint)_handle;
            }
        }

        public bool IsValid
        {
            get
            {
                return _handle != null;
            }
        }

        public Dimension Dimension
        {
            get
            {
                return (Dimension)NativeApi.Table->levelDimension(_handle);
            }
        }

        public string Name
        {
            get
            {
                return NativeApi.Text(NativeApi.Table->levelName(_handle));
            }
        }

        public long Time
        {
            get
            {
                return NativeApi.Table->levelTime(_handle);
            }
            set
            {
                NativeApi.Table->levelSetTime(_handle, value);
            }
        }

        public bool IsRaining
        {
            get
            {
                return NativeApi.Table->levelIsRaining(_handle) != 0;
            }
            set
            {
                NativeApi.Table->levelSetRaining(_handle, value ? 1 : 0);
            }
        }

        public bool IsThundering
        {
            get
            {
                return NativeApi.Table->levelIsThundering(_handle) != 0;
            }
            set
            {
                NativeApi.Table->levelSetThundering(_handle, value ? 1 : 0);
            }
        }

        public Vec3 SpawnPosition
        {
            get
            {
                return Vec3.From(NativeApi.Table->levelSpawnPosition(_handle));
            }
        }

        public IReadOnlyList<Entity> Entities
        {
            get
            {
                uint count = NativeApi.Table->levelEntityCount(_handle);
                List<Entity> entities = new List<Entity>((int)count);
                for (uint index = 0; index < count; index++)
                {
                    entities.Add(new Entity(NativeApi.Table->levelEntity(_handle, index)));
                }

                return entities;
            }
        }

        internal FalconLevel* Raw
        {
            get
            {
                return _handle;
            }
        }

        public static bool operator ==(Level? left, Level? right)
        {
            if (left is null)
            {
                return right is null;
            }

            return left.Equals(right);
        }

        public static bool operator !=(Level? left, Level? right)
        {
            return !(left == right);
        }

        public Block GetBlock(BlockPos position)
        {
            return new Block(position, BlockName(position), BlockStates(position));
        }

        public string BlockName(BlockPos position)
        {
            return NativeApi.Text(NativeApi.Table->levelGetBlock(_handle, position.ToNative()));
        }

        public string BlockStates(BlockPos position)
        {
            return NativeApi.Text(NativeApi.Table->levelGetBlockStates(_handle, position.ToNative()));
        }

        public bool SetBlock(BlockPos position, string name, string statesJson = "")
        {
            fixed (byte* nameText = NativeApi.Utf8(name))
            fixed (byte* statesText = NativeApi.Nullable(statesJson))
            {
                return NativeApi.Table->levelSetBlock(_handle, position.ToNative(), nameText, statesText) != 0;
            }
        }

        public bool BreakBlock(BlockPos position, bool dropItems = true)
        {
            return NativeApi.Table->levelBreakBlock(_handle, position.ToNative(), dropItems ? 1 : 0) != 0;
        }

        public bool IsChunkLoaded(int chunkX, int chunkZ)
        {
            return NativeApi.Table->levelIsChunkLoaded(_handle, chunkX, chunkZ) != 0;
        }

        public int HighestBlockY(int x, int z)
        {
            return NativeApi.Table->levelHighestBlockY(_handle, x, z);
        }

        public Entity? SpawnEntity(string identifier, Vec3 position)
        {
            fixed (byte* text = NativeApi.Utf8(identifier))
            {
                return Entity.Wrap(NativeApi.Table->levelSpawnEntity(_handle, text, position.ToNative()));
            }
        }

        public void DropItem(Vec3 position, Item item)
        {
            NativeApi.Table->levelDropItem(_handle, position.ToNative(), item.Raw);
        }

        public void StrikeLightning(Vec3 position)
        {
            NativeApi.Table->levelStrikeLightning(_handle, position.ToNative());
        }

        public void CreateExplosion(Vec3 position, float power, bool breakBlocks = true)
        {
            NativeApi.Table->levelCreateExplosion(_handle, position.ToNative(), power, breakBlocks ? 1 : 0);
        }

        public bool Equals(Level? other)
        {
            return other is not null && other._handle == _handle;
        }

        public override bool Equals(object? other)
        {
            return Equals(other as Level);
        }

        public override int GetHashCode()
        {
            return ((nint)_handle).GetHashCode();
        }

        internal static Level? Wrap(FalconLevel* handle)
        {
            if (handle == null)
            {
                return null;
            }

            return new Level(handle);
        }
    }
}

File: java/src/main/java/falcon/api/Level.java
package falcon.api;

import falcon.api.internal.Interop;
import falcon.api.internal.NativeApi;
import falcon.api.internal.PluginContext;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public final class Level {
    private final PluginContext mContext;
    private final MemorySegment mHandle;

    public Level(PluginContext context, MemorySegment handle) {
        mContext = context;
        mHandle = handle == null ? MemorySegment.NULL : handle;
    }

    public boolean valid() {
        return !Interop.isNull(mHandle);
    }

    public Dimension dimension() {
        return Dimension.of(api().levelDimension(mHandle));
    }

    public String name() {
        return Interop.string(api().levelName(mHandle));
    }

    public Block getBlock(BlockPos position) {
        return new Block(position, blockName(position), blockStates(position));
    }

    public String blockName(BlockPos position) {
        try (Arena arena = Arena.ofConfined()) {
            return Interop.string(api().levelGetBlock(mHandle, Interop.blockPos(arena, position)));
        }
    }

    public String blockStates(BlockPos position) {
        try (Arena arena = Arena.ofConfined()) {
            return Interop.string(api().levelGetBlockStates(mHandle, Interop.blockPos(arena, position)));
        }
    }

    public boolean setBlock(BlockPos position, String name) {
        return setBlock(position, name, "");
    }

    public boolean setBlock(BlockPos position, String name, String statesJson) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment target = Interop.blockPos(arena, position);
            MemorySegment states = Interop.nullableText(arena, statesJson);
            return api().levelSetBlock(mHandle, target, Interop.text(arena, name), states) != 0;
        }
    }

    public boolean breakBlock(BlockPos position) {
        return breakBlock(position, true);
    }

    public boolean breakBlock(BlockPos position, boolean dropItems) {
        try (Arena arena = Arena.ofConfined()) {
            return api().levelBreakBlock(mHandle, Interop.blockPos(arena, position), dropItems ? 1 : 0) != 0;
        }
    }

    public boolean isChunkLoaded(int chunkX, int chunkZ) {
        return api().levelIsChunkLoaded(mHandle, chunkX, chunkZ) != 0;
    }

    public int highestBlockY(int x, int z) {
        return api().levelHighestBlockY(mHandle, x, z);
    }

    public long time() {
        return api().levelTime(mHandle);
    }

    public void setTime(long time) {
        api().levelSetTime(mHandle, time);
    }

    public boolean isRaining() {
        return api().levelIsRaining(mHandle) != 0;
    }

    public void setRaining(boolean raining) {
        api().levelSetRaining(mHandle, raining ? 1 : 0);
    }

    public boolean isThundering() {
        return api().levelIsThundering(mHandle) != 0;
    }

    public void setThundering(boolean thundering) {
        api().levelSetThundering(mHandle, thundering ? 1 : 0);
    }

    public Vec3 spawnPosition() {
        try (Arena arena = Arena.ofConfined()) {
            return Interop.vec3(api().levelSpawnPosition(arena, mHandle));
        }
    }

    public Optional<Entity> spawnEntity(String identifier, Vec3 position) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment target = Interop.vec3(arena, position);
            MemorySegment entity = api().levelSpawnEntity(mHandle, Interop.text(arena, identifier), target);
            if (Interop.isNull(entity)) {
                return Optional.empty();
            }
            return Optional.of(new Entity(mContext, entity));
        }
    }

    public void dropItem(Vec3 position, Item item) {
        try (Arena arena = Arena.ofConfined()) {
            api().levelDropItem(mHandle, Interop.vec3(arena, position), item.handle());
        }
    }

    public void strikeLightning(Vec3 position) {
        try (Arena arena = Arena.ofConfined()) {
            api().levelStrikeLightning(mHandle, Interop.vec3(arena, position));
        }
    }

    public void createExplosion(Vec3 position, float power) {
        createExplosion(position, power, true);
    }

    public void createExplosion(Vec3 position, float power, boolean breakBlocks) {
        try (Arena arena = Arena.ofConfined()) {
            api().levelCreateExplosion(mHandle, Interop.vec3(arena, position), power, breakBlocks ? 1 : 0);
        }
    }

    public List<Entity> entities() {
        int count = api().levelEntityCount(mHandle);
        List<Entity> result = new ArrayList<>(count);
        for (int index = 0; index < count; index++) {
            result.add(new Entity(mContext, api().levelEntity(mHandle, index)));
        }
        return result;
    }

    public MemorySegment handle() {
        return mHandle;
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof Level level && level.mHandle.address() == mHandle.address();
    }

    @Override
    public int hashCode() {
        return Long.hashCode(mHandle.address());
    }

    private static NativeApi api() {
        return Interop.api();
    }
}
